### PR TITLE
Move comment moderation into a new Servlet

### PR DESCRIPTION
**Pull request by n-wach**
This will allow verifying that a comment is ok to post before submitting it (similar to the nickname check).

### CODE REVIEW

**Comment by n-wach** [comment]
https://screenshot.googleplex.com/gFzHmLPKgZ1

File: portfolio/src/main/java/com/google/sps/Moderator.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    try {
+      Secrets secrets = Secrets.getSecrets();
+      perspective = PerspectiveAPI.create(secrets.getPerspectivesApiKey());
+    } catch (IOException e) {
+      // This is called from a static context so throwing an unchecked RuntimeException is our
+      // best option.
+      throw new RuntimeException("Unable to initialize the Perseistence API for comment" +
+          "moderation");
+    }
+  }
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/Moderator.java:51]: Perhaps instead of this, we can handle this in getModerator().
```
@Nullable
public static Moderator getModerator(){
 if(instance == null){
  try {
      Secrets secrets = Secrets.getSecrets();
      perspective = PerspectiveAPI.create(secrets.getPerspectivesApiKey());
      instance = new Moderator();
    } catch (IOException e) {
      // log something
      instance =  null;
    }
}

return instance;
}
```

File: portfolio/src/main/java/com/google/sps/Moderator.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  public static Moderator getModerator() {
+    return INSTANCE;
+  }
+
+  /**
+   * Returns {@code true} if a comment has acceptable levels of {@code TOXICITY}.
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/Moderator.java:58]: nit:
```suggestion
   * Returns {@code true} if a comment has acceptable levels of {@link  AttributeType.TOXICITY}.
```

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  public void init() throws ServletException {
+    auth = Authenticator.getAuthenticator();
+    moderator = Moderator.getModerator();
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Only logged in users can submit comments, so this prevents abuse.
+    if (!auth.isUserLoggedIn()) {
+      response.sendError(403, "Not logged in");
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:51]: Maybe a better error message like, "Looks like you're not logged in"?

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -129,6 +130,42 @@ function makeSubmission(topic) {
...
+
+    const request = await fetch(url);
+    const moderationResponse = await request.json();
+
+    if (!moderationResponse.isAcceptable) {
+      submit.disabled = true;
+      const toxicity = moderationResponse.analysis.attributeScores.TOXICITY;
+      const score = toxicity.summaryScore.value;
+      const percent = (score * 100).toFixed();
+      submit.value = `Try being a little nicer.  (TOXICITY at ${percent}%)`;
```
Comment (f-camp) [inline portfolio/src/main/webapp/script.js:145]: How about 'Try being a little nicer. Looks like your comment might not be so great' ?
I don't know if the user would understand the TOXICITY percentage or why it was measured that way. Might be best to just have a comment.

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -129,6 +130,42 @@ function makeSubmission(topic) {
...
+    if (!onCooldown) {
+      checkModeration(event);
+      onCooldown = true;
+      clearTimeout(resetCooldown);
+      resetCooldown = setTimeout(() => {
+        onCooldown = false;
+        checkModeration(event);
+      }, 3000);
+    }
+  });
```
Comment (f-camp) [inline portfolio/src/main/webapp/script.js:167]: Mmh, what's this piece of code doing?
It looks like we're trying to make sure that we moderate the comment before submitting?
maybe we can rename onCooldown to a more intuitive variable name and a bit of comment on the listener to explain what's going on.

Comment (n-wach) [reply portfolio/src/main/webapp/script.js:167]: I'm only allowed to query the perspectives API 60 times a minute.  Before I was sending a query whenever the comment text changed, but that was wayyy too often.  This code just makes sure we wait at least 3 seconds between queries.  You're right though, it's pretty messy.  I'll clean it up!

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:145]: Yeah I'll change that.  I think I'll show the TOXICITY (and an explanation) but only when you hover over the error, it's fun to experiment with...

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java (hunk 3 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:51]: The only way this error will appear is if someone is nefariously sending their own request to /moderation.  The comment box (which does the moderation queries) will only appear if someone is logged in.  A user will never see this error.  This is consistent with the errors in the comment and nickname servlets, so I'm gonna keep it as is, if that's alright

File: portfolio/src/main/java/com/google/sps/Moderator.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import au.com.origma.perspectiveapi.v1alpha1.models.AnalyzeCommentResponse;
+import au.com.origma.perspectiveapi.v1alpha1.models.AttributeScore;
+import au.com.origma.perspectiveapi.v1alpha1.models.AttributeType;
+import au.com.origma.perspectiveapi.v1alpha1.models.ContentType;
+import java.io.IOException;
+
+import javax.annotation.Nullable;
+
+/**
+ * Uses the Perspectives API to moderate comments.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/Moderator.java:29]: I would rephrase as: Moderates comments using the Perspectives API.

The main action is moderating not using an api.

File: portfolio/src/main/java/com/google/sps/Moderator.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+
+  /**
+   * Returns an {@code AnalyzeCommentResponse} from the Perspective API analyzing {@code content}.
+   *
+   * <p>This will return {@code null} if {@code content} is an empty string.
+   */
+  @Nullable
+  public AnalyzeCommentResponse analyze(String content) {
+    return perspective.analyze(new AnalyzeCommentRequest.Builder()
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/Moderator.java:74]: you can do: return Optional.ofNullable(...);

That way you do not have to return Nullable.

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+   * <p>This asks {@code Moderator} to send a request to the Perspectives API and judges the
+   * response using {@link com.google.sps.Moderator.isAcceptableComment}. This returns
+   * {@code null} if the moderator returns {@code null} when attempting to evaluate the comment.
+   * This can happen if the API encouters an error (such as when evaluating an empty comment).
+   */
+  @Nullable
+  public ModerationResponse moderate(String comment) {
+    AnalyzeCommentResponse analysis = moderator.analyze(comment);
+    if (analysis == null) {
+      return null;
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:87]: Can you return Optional.empty()

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    final ModerationResponse response;
+    if (moderator.isAcceptableComment(analysis)) {
+      response = ModerationResponse.acceptable(analysis);
+    } else {
+      response = ModerationResponse.unacceptable(analysis);
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:94]: for here instead of storing in response variable, just return the response. The benefit of this is you do not need an 'else' statement.

Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:94]: done

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java (hunk 8 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:87]: I'll switch to using optional here

File: portfolio/src/main/java/com/google/sps/Moderator.java (hunk 7 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/Moderator.java:74]: I'll switch to using optional here

File: portfolio/src/main/java/com/google/sps/Moderator.java (hunk 2 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/Moderator.java:58]: done

File: portfolio/src/main/java/com/google/sps/Moderator.java (hunk 6 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/Moderator.java:29]: Done

File: portfolio/src/main/java/com/google/sps/Moderator.java (hunk 1 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/Moderator.java:51]: I decided to just have the getModerator throw the exception, so I can catch it and throw ServletExceptions in the init methods of servlets that need a moderator.

Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/Moderator.java:51]: Let me know if it looks good

File: portfolio/src/main/webapp/script.js (hunk 5 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:167]: Done

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:145]: Done

File: portfolio/src/main/java/com/google/sps/Moderator.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+  public static Moderator getModerator() throws IOException {
+    if (instance == null) {
+      Secrets secrets = Secrets.getSecrets();
+      // Creating the API may throw an IOException.
+      PerspectiveAPI perspective = PerspectiveAPI.create(secrets.getPerspectivesApiKey());
+      instance = new Moderator(perspective);
+    }
+
+    // At this point instance cannot be null.
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/Moderator.java:53]:
```suggestion
```
This is obvious from the code

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java (hunk 3 above)
Comment (f-camp) [reply portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:55]: It sounds like this is a system error.  "User not logged in" would sound better, "Not logged in" is vague still.

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    String content = request.getParameter(PARAM_CONTENT);
+    if (content == null || content.isEmpty()) {
+      response.sendError(400, "No content specified");
+      return;
+    }
+
+    Optional<ModerationResponse> r = moderate(content);
+
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:66]: Remove new line please.
r is not an intuitive variable name. Please use update to something else, maybe moderationResponse?

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+  public Optional<ModerationResponse> moderate(String comment) {
+    Optional<AnalyzeCommentResponse> analysis = moderator.analyze(comment);
+    if (analysis.isEmpty()) {
+      return Optional.empty();
+    }
+
+    if (moderator.isAcceptableComment(analysis)) {
+      return Optional.of(ModerationResponse.acceptable(analysis.get()));
+    }
+
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:97]: Unnecessary new line please remove.

File: portfolio/src/main/webapp/script.js [hunk 13]
```diff
@@ -129,6 +131,44 @@ function makeSubmission(topic) {
   as.innerText = `as ${USER_INFO.user.value.nickname} or `;
   form.appendChild(as);
 
+  async function checkModeration(event) {
+    const url = new URL(MODERATION_ENDPOINT, window.location.origin);
+    url.searchParams.append(PARAM_CONTENT, event.target.value);
+
+    const request = await fetch(url);
+    const moderationResponse = await request.json();
+
```
Comment (f-camp) [inline portfolio/src/main/webapp/script.js:140]: Unnecessary new line please remove.

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -129,6 +131,44 @@ function makeSubmission(topic) {
...
+    if (!moderationResponse.isAcceptable) {
+      submit.disabled = true;
+      const toxicity = moderationResponse.analysis.attributeScores.TOXICITY;
+      const score = toxicity.summaryScore.value;
+      const percent = (score * 100).toFixed();
+      submit.value = 'Try being nicer. Looks like your comment ' +
+          'might be a little mean.';
+      submit.title = (`Your comment's toxicity score is ${percent}%, which is` +
+          'above the 80% threshold.');
+
```
Comment (f-camp) [inline portfolio/src/main/webapp/script.js:150]: Unnecessary new line please remove.

File: portfolio/src/main/webapp/script.js [hunk 15]
```diff
@@ -33,11 +33,13 @@ const PARAM_NICKNAME = 'nickname';
 const COMMENTS_ENDPOINT = '/comments';
 const USER_ENDPOINT = '/user';
 const NICKNAME_ENPOINT = '/nickname';
+const MODERATION_ENDPOINT = '/moderation'
 
 const SELECTOR_THREAD = '.comment-thread';
 const SELECTOR_PROJECT = '.project';
 
 const AUTHOR_ANONYMOUS = 'anonymous';
+const MODERATOR_QUERY_COOLDOWN = 3000;
```
Comment (f-camp) [inline portfolio/src/main/webapp/script.js:42]: What unit is this millis, nanos?  We usually indicate this in the variable name eg. MODERATOR_QUERY_COOLDOWN_MILLIS.
Please kindly update to have unit.

File: portfolio/src/main/java/com/google/sps/Moderator.java (hunk 10 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/Moderator.java:53]: removed

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java (hunk 12 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:97]: fixed

File: portfolio/src/main/webapp/script.js (hunk 14 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:150]: Done

File: portfolio/src/main/webapp/script.js (hunk 13 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:140]: Done

File: portfolio/src/main/webapp/script.js (hunk 15 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:42]: used _MS

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java (hunk 11 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:66]: Fixed

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java (hunk 3 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java]: Done

File: portfolio/src/main/java/com/google/sps/Moderator.java [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+      Secrets secrets = Secrets.getSecrets();
+      // Creating the API may throw an IOException.
+      PerspectiveAPI perspective = PerspectiveAPI.create(secrets.getPerspectivesApiKey());
+      instance = new Moderator(perspective);
+    }
+    return instance;
+  }
+
+  /**
+   * Returns {@code true} if a comment has acceptable levels of {@link AttributeType.TOXICITY}.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/Moderator.java:56]:
```suggestion
   * Returns {@code true} if a comment has an acceptable level of {@link AttributeType.TOXICITY}.
```

File: portfolio/src/main/java/com/google/sps/Moderator.java [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+   */
+  public boolean isAcceptableComment(Optional<AnalyzeCommentResponse> analysis) {
+    if (analysis.isEmpty()) {
+      return false;
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/Moderator.java:60]: I would add a comment here explaining why you decided to return false if the analysis is empty.

Comment (billie-cat) [reply portfolio/src/main/java/com/google/sps/Moderator.java:60]: Seems it is empty if the string is empty. Wouldn't that make the comment acceptable?

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 18]
```diff
@@ -35,12 +28,14 @@
...
 import com.google.common.collect.ImmutableList;
 import com.google.gson.Gson;
-import com.google.sps.Secrets;
+import com.google.sps.Moderator;
 import com.google.sps.auth.Authenticator;
 import com.google.sps.data.Comment;
 import com.google.sps.data.CommentThread;
 import com.google.sps.data.User;
 
+import au.com.origma.perspectiveapi.v1alpha1.models.AnalyzeCommentResponse;
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:37]: These should be alphabetized and au comes before com.

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+import java.io.IOException;
+import java.util.Optional;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.ServletException;
+
+/**
+ * Uses the Perspectives API to provide moderation info for comments.  This is used by the client
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:31]:
```suggestion
 * Provides moderation info for comments using the Perspectives API.  This is used by the client
```

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+
+    Gson gson = new Gson();
+    String threadJson = gson.toJson(moderationResponse.get());
+    response.setContentType("application/json;");
+    response.getWriter().println(threadJson);
+  }
+
+
+  /**
+   * Returns a {@code ModerationResponse} for a comment.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:79]:
```suggestion
   * Returns a {@code ModerationResponse} for {@code comment}.
```

File: portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+   *
+   * <p>This asks {@code Moderator} to send a request to the Perspectives API and judges the
+   * response using {@link com.google.sps.Moderator.isAcceptableComment}. This returns
+   * {@code Optional.empty()} if the moderator returns {@code Optional.empty()} when attempting
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:83]: Special return values should be part of the first JavaDoc sentence as it is important information. I would rephrase this entire JavaDoc as this:

Returns {@code ModerationResponse} for {@code comment}. Returns {@code Optional.empty()} if an error occurred during the analysis of {@code comment}.

<p>Asks {@code Moderator} to send a request a request to the Perspectives API and judges the response using {@link com.google.sps.Moderator.isAcceptableComment}

Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/ModerationServlet.java:83]: Done
